design-docs/opmodes.md

```diff
@@ -0,0 +1,516 @@
+- `int GetSelectedTeleoperatedOpMode()` - gets the opmode ID currently selected for teleoperated; valid when the robot is disabled; returns 0 if unknown or no selection
+- `int GetSelectedTestOpMode()` - gets the opmode ID currently selected for test; valid when the robot is disabled; returns 0 if unknown or no selection
+
+## DS Protocol
+
+The list of opmodes (and details thereof) is communicated from the Robot to the DS via the tagged TCP link.
+
+As in the current FRC protocol, the enabled state and teleop/auto/test selection are sent as part of the UDP control word from the DS to the Robot. Additionally, the selected opmodes for auto, teleop, and test are sent as part of every UDP packet.  This is done via a 64-bit hash of each opmode's name.  Hash collisions are extremely low probability given the low quantity of opmodes, but this can be checked on the robot side when generating the opmode list, and spaces appended as necessary to the provided names to uniquify the hashes.
+
+# Migration from 2025 FRC (WPILib)
+
+Key differences from 2025 FRC:
+- The per-opmode overrideable functions in Robot are replaced with separate annotated per-opmode classes (for non-command-based)
+- `SendableChooser` is no longer required for selecting autonomous routines; instead this functionality is integrated into opmodes, as users can create/register multiple autonomous opmodes classes, and it's been extended to support multiple teleoperated and multiple test opmodes
+- Selection of autonomous opmodes is integrated into the DS instead of being performed by the dashboard
+- OpModes are no longer limited to just timed (periodic); a mix of different opmode functionality/approaches across different robot opmodes is possible (e.g. teleop can be periodic and autonomous linear)
+- Linear opmodes are now available; these are improved versions of the old SimpleRobot/SampleRobot
+
+# Migration from 2025 FTC (FTC SDK)
+
+Key differences from 2025 FTC:
+- There is no hardware map.  Hardware objects are instead directly instantiated inside a top-level Robot class.  This Robot class is provided as a parameter to the opmode constructor.
+- Init is integrated into the opmode constructor.  There is no equivalent to the the enabled init step; the opmode constructor is run as soon as the opmode is selected on the DS, while the robot is still disabled.  Match periods will start as soon as the robot is enabled.
+- There is no `@Disabled` annotation for opmodes; the opmode annotation can simply be commented out to achieve the same result.
+
+# Drawbacks
+
+# Alternatives
+
+Use SendableChooser for more opmodes (teleop and test as well as autonomous); downsides of this:
+- Doesn't allow flexibility of mixing linear and periodic opmodes
+- Harder to use than annotations (generic class)
+- No grouping, no filtering by opmode (although these could be added)
+
+# Trades
+
+- Overall naming: mode? routine? opmode?
```

Comment:
Mode is definitely a standard nomenclature (e.g "What auto modes does your team have?"). 'Teleop OpMode' also doesn't roll off the tongue, and will likely be shortened to 'teleop mode' in the vernacular.

Reply:
We have literally gone back and forth several times on what to call this.  Just calling it “mode” has potential confusion as well.